Source/DiabloUI/diabloui.cpp

```diff
@@ -226,15 +225,13 @@ void UiFocusPageDown()
 
 void SelheroCatToName(char *inBuf, char *outBuf, int cnt)
 {
-	std::string output = utf8_to_latin1(inBuf);
-	strncat(outBuf, output.c_str(), cnt - strlen(outBuf));
+	strncat(outBuf, inBuf, cnt - strlen(outBuf));
 }
 
 #ifdef __vita__
 void selhero_SetName(char *in_buf, char *out_buf, int cnt)
 {
-	std::string output = utf8_to_latin1(in_buf);
-	strncpy(out_buf, output.c_str(), cnt);
+	strncpy(out_buf, in_buf, cnt);
 }
 #endif

```

Comment:
You get a decode error and the string stops rendering at that point. So usually it should be acceptable.

Reply:
It seems that backspace will screw up multi-byte characters. To make this hero, I typed `Ā<backspace>erika`.

![image](https://user-images.githubusercontent.com/9203145/134281368-97dcb1a9-bf49-4ec0-8507-b15b3745a30f.png)

Reply:
Right that would cause a decoding issue with the very first character, since it only deletes half of it. 

Shouldn't be to hard to fix.

Reply:
Actually, backspace seems to work fine for the in-game chat so I intentionally didn't mention it.

Reply:
It looks like backspace works for the in-game chat because we're using the return value from `DrawString()` to force a null character onto the "end" of it.

https://github.com/diasurgical/devilutionX/blob/3426a3d76f11a5973c50915eb3fa3185508be7bf/Source/control.cpp#L1689-L1690

This works because the rendering stops when it encounters an error so the null character overwrites the broken character. Incidentally, this is also the only place in the whole codebase where the return value of `DrawString()` is used.

Reply:
Yeah I did look in to getting rid of this, but it's not super simple to do, we would need a function that does almost the same to calculate the cut off point. Maybe something could be done via splitting DrawString() up in to sub routines, but atm. I don't think it's worth it.

Reply:
Based on what I know of UTF-8, I think backspace could be done with...

```c++
std::size_t index = strlen(TalkMessage) - 1;
while (index > 0 && (TalkMessage[index] & 0xC0) == 0x80)
	index--;
if (index >= 0)
	TalkMessage[index] = '\0';
```

This assumes there are no broken characters in the string. If there are broken characters, then you might see some weird behavior, but nothing catastrophic.

Reply:
In case it wasn't clear my previous comment was in regards to `DrawString()` returning a count for what will fit in a box. A potential solution for that would be to do a `GetLineWidt(TalkMessage) < 3 * width` check in `control_new_text()` and prevent the text from being added in the first place.

As for a back space function i simply decode the string and return the position of the last found symbol (see latest commit). Your method should work as well and is a bit simpler, but might be a bit more cryptic to read, let me know if you think it would be better to use it.

Reply:
Ahh, I get it now. I had been thinking that surely backspace wasn't the only reason for using that return value.

As for how you handled backspace, your function should work better than my suggestion.